crypto/bsslmlkem/mlkem768.c

```diff
@@ -0,0 +1,1140 @@
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* Copyright (c) 2024, Google Inc. */
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <openssl/rand.h>
+#include <openssl/err.h>
+#include <crypto/mlkem.h>
+#include <internal/sha3.h>
+#include <internal/constant_time.h>
+#include <internal/common.h>
+#ifndef NDEBUG
+# include <stdio.h>
+#endif
+
+#ifndef OPENSSL_NO_MLKEM
+
+/* Constants that are common across all sizes. */
+# define DEGREE 256
+static const size_t kBarrettMultiplier = 5039;
+static const unsigned kBarrettShift = 24;
+static const uint16_t kPrime = 3329;
+static const int kLog2Prime = 12;
+static const uint16_t kHalfPrime = (/* kPrime= */ 3329 - 1) / 2;
+
+/*
+ * kInverseDegree is 128^-1 mod 3329; 128 because kPrime does not have a 512th
+ * root of unity.
+ */
+static const uint16_t kInverseDegree = 3303;
+
+/* Rank-specific constants. */
+# define RANK768 3
+static const int kDU768 = 10;
+static const int kDV768 = 4;
+# define RANK1024 4
+static const int kDU1024 = 11;
+static const int kDV1024 = 5;
+
+static size_t compressed_vector_size(int rank)
+{
+    return (rank == RANK768 ? kDU768 : kDU1024) * (size_t)rank *
+        DEGREE / 8;
+}
+
+static size_t ciphertext_size(int rank)
+{
+    return compressed_vector_size(rank) +
+        (rank == RANK768 ? kDV768 : kDV1024) * DEGREE / 8;
+}
+
+typedef struct scalar {
+    /* On every function entry and exit, 0 <= c < kPrime. */
+    uint16_t c[DEGREE];
+} scalar;
+
+/* TODO(ML-KEM): possibly rename vector768 to allow for other algs */
+typedef struct vector {
+    scalar v[RANK768];
+} vector;
+
+/* TODO(ML-KEM): possibly rename matrix768 to allow for other algs */
+typedef struct matrix {
+    scalar v[RANK768][RANK768];
+} matrix;
+
+typedef struct public_key_RANK768
+{
+    vector t;
+    uint8_t rho[32];
+    uint8_t public_key_hash[32];
+    matrix m;
+} public_key_RANK768;
+
+typedef struct private_key_RANK768
+{
+    struct public_key_RANK768 pub;
+    vector s;
+    uint8_t fo_failure_secret[32];
+} private_key_RANK768;
+
+static size_t encoded_vector_size(int rank)
+{
+    return (kLog2Prime * DEGREE / 8) * (size_t)rank;
+}
+
+static size_t encoded_public_key_size(int rank)
+{
+    return encoded_vector_size(rank) + /* sizeof(rho)= */ 32;
+}
+
+/*
+ * MLKEM_ENCAP_ENTROPY is the number of bytes of uniformly random entropy
+ * necessary to encapsulate a secret. The entropy will be leaked to the
+ * decapsulating party.
+ */
+# define MLKEM_ENCAP_ENTROPY 32
+
+/* MD&XOF handles */
+
+/* TODO(ML-KEM): enhance as per https://github.com/openssl/private/issues/700 */
+
+static EVP_MD *shake128_cache = NULL;
+static EVP_MD *shake256_cache = NULL;
+static EVP_MD *sha3_256_cache = NULL;
+static EVP_MD *sha3_512_cache = NULL;
```

Comment:
One possibility is to have some MLKEM_CACHE structure to cache these - and then pass that in as a parameter to the relevant functions. It could then be cached at the PROVCTX level which avoids lots of issues about when to create these, and what happens if the provider gets unloaded, etc.

Reply:
See discussion at https://github.com/baentsch/openssl/pull/2#discussion_r1827344176 - already marked TODO(ML-KEM).